packages/jsii-kernel/lib/kernel.ts

```diff
@@ -923,12 +924,17 @@ export class Kernel {
         // so the client receives a real object.
         if (typeof(v) === 'object' && targetType && spec.isNamedTypeReference(targetType)) {
             this._debug('coalescing to', targetType);
-            const newObjRef = this._create({ fqn: targetType.fqn });
-            const newObj = this._findObject(newObjRef);
-            for (const k of Object.keys(v)) {
-                newObj[k] = v[k];
-            }
-            return newObjRef;
+            const proxies = v[PROXIES_PROP] = v[PROXIES_PROP] || {};
```

Comment:
Why not just call `this._createObjref(target, targetType.fqn)`. Wouldn't that achieve the same goal?

Reply:
The same object literal may be returned under different type identities, so you cannot issue an object proxy that has only one type & expect it to "always work".

Reply:
Got it, so the behavior of `_createObjref` doesn't make sense in that respect because it adds FQN_PROP and OBJID_PROP to the target, and as you say, the same object can represent two different FQNs. I am wondering if in this case it should still represent the same OBJID, which also contains the FQN...

Reply:
What FQN, though? I like the idea of being able to return a single OBJID for the object literal (because it would make it trivial to implement identity checks, or even de-duplication on the front-ends)... However, consider the following (valid) typescript:

```ts
export class TrickyType {
  private readonly _magic: TypeA & TypeB = {};

  public get magicA(): TypeA {
    return this._magic;
  }

  public get magicB(): TypeB {
    return this._magic;
  }
}

export interface TypeA {
  foo?: string;
}

export interface TypeB {
  bar?: string;
}
```

Now, if you want `magicA` and `magicB` to return the same object reference, which of `TypeA` and `TypeB` do you tag in the `ObjRef`, knowing front-end languages will use this to determine which class to make a proxy instance for?

Reply:
The path I've taken here has a weaker guarantee that if you call `magicA` several times, you'll get the same object reference each time... But `magicB` will be another. We can make it be a single reference, however it'd require (way) more significant changes (basically, the `ObjRef` shape will need to change so the type information is modeled separately).

Reply:
I am okay with that. The current implementation returned a different object reference in every call, so that's definitely an improvement. Worth documenting in code though

Reply:
I've added comments... let me know if not clear enough (or too verbose? :D)